File: 백준/Gold/3079. 입국심사/입국심사.py
def can_process(time, times, m):
    total = 0
    for t in times:
        total += time // t
    return total >= m


n,m = map(int,input().split())
times = [int(input()) for _ in range(n)]

left = 1
right = max(times) * m

answer = right

while left <= right:
    mid = (left + right) // 2
    if can_process(mid, times, m):
        answer = mid
        right = mid - 1
    else:
        left = mid + 1
print(answer)

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;

class Solution {
    public long solution(int n, int[] times) {
        long left = 1;
        long right = (long) Arrays.stream(times).max().getAsInt() * n;
        long min = 0;
        
        while (left <= right){
            long mid = (left + right) / 2;
            
            long available = 0;
            
            for (int time: times){
                available += (mid / time);
            }
            
            if (available >= n){
                // 충분한 경우
                min = mid;
                right = mid - 1;
            }else{
                left = mid + 1;
            }
        }
        return min;
    }
}